File: core/src/net/artux/pda/map/engine/systems/SpawnSystem.java
package net.artux.pda.map.engine.systems;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.systems.IteratingSystem;
import com.badlogic.ashley.utils.ImmutableArray;
import com.badlogic.gdx.Gdx;

import net.artux.pda.map.DataRepository;
import net.artux.pda.map.di.scope.PerGameMap;
import net.artux.pda.map.engine.components.GroupComponent;
import net.artux.pda.map.engine.components.PositionComponent;
import net.artux.pda.map.engine.components.SpawnComponent;
import net.artux.pda.map.engine.components.VelocityComponent;
import net.artux.pda.map.engine.components.VisionComponent;

import java.util.List;

import javax.inject.Inject;

@PerGameMap
public class SpawnSystem extends IteratingSystem {

    private ImmutableArray<Entity> spawns;
    private ImmutableArray<Entity> groupEntities;
    private final DataRepository dataRepository;

    private final ComponentMapper<SpawnComponent> sm = ComponentMapper.getFor(SpawnComponent.class);
    private final ComponentMapper<PositionComponent> pm = ComponentMapper.getFor(PositionComponent.class);
    private final ComponentMapper<GroupComponent> gm = ComponentMapper.getFor(GroupComponent.class);

    @Inject
    public SpawnSystem(DataRepository dataRepository) {
        super(Family.all(VisionComponent.class, PositionComponent.class, VelocityComponent.class).get());
        this.dataRepository = dataRepository;
    }

    @Override
    public void addedToEngine(Engine engine) {
        super.addedToEngine(engine);
        spawns = engine.getEntitiesFor(Family.all(SpawnComponent.class, PositionComponent.class).exclude(VelocityComponent.class).get());
        groupEntities = engine.getEntitiesFor(Family.all(GroupComponent.class, PositionComponent.class).get());
    }

    @Override
    public void update(float deltaTime) {
        super.update(deltaTime);
        for (int i = 0; i < spawns.size(); i++) {
            SpawnComponent spawnComponent = sm.get(spawns.get(i));
            PositionComponent spawnPosition = pm.get(spawns.get(i));

            List<Entity> entities = spawnComponent.getGroupComponent().getEntities();
            entities.removeIf(e -> getEntities().indexOf(e, true) < 0);
            if (spawnComponent.isEmpty()) {
                if (!spawnComponent.isActionsDone()) {
                    Gdx.app.log("Spawn actions", "Actions sent");
                    dataRepository.applyActions(spawnComponent.getSpawnModel().getActions());
                    spawnComponent.setActionsDone(true);
                }

                GroupComponent minDstGroup = null;
                float minDst = 50;
                for (Entity entity : groupEntities) {
                    GroupComponent group = gm.get(entity);
                    if (!(group.getTargeting() instanceof SpawnComponent)) {
                        float dst = group.getCenterPoint().dst(spawnPosition);
                        if (minDstGroup == null
                                || dst < minDst) {
                            minDstGroup = group;
                            minDst = dst;
                        }
                        spawnComponent.setGroup(group);
                    }
                }

            }
        }

    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {

    }

    public Entity getEmptySpawn() {
        for (Entity entity : spawns) {
            SpawnComponent spawnComponent = sm.get(entity);
            if (spawnComponent.isEmpty()) {
                return entity;
            }
        }
        return null;
    }

    public Entity getRandomSpawn() {
        return spawns.random();
    }
}

File: core/src/main/java/net/artux/pda/map/ecs/takeover/SpawnSystem.kt
package net.artux.pda.map.ecs.takeover

import com.badlogic.ashley.core.*
import com.badlogic.ashley.utils.ImmutableArray
import com.badlogic.gdx.Gdx
import net.artux.pda.map.repository.DataRepository
import net.artux.pda.map.ecs.physics.BodyComponent
import net.artux.pda.map.ecs.ai.StalkerGroup
import net.artux.pda.map.ecs.interactive.PassivityComponent
import net.artux.pda.map.ecs.vision.VisionComponent
import net.artux.pda.map.ecs.interactive.map.SpawnComponent
import net.artux.pda.map.ecs.systems.BaseSystem
import net.artux.pda.map.ecs.creation.EntityProcessorSystem
import net.artux.pda.map.controller.notification.NotificationController
import net.artux.pda.map.di.scope.PerGameMap
import javax.inject.Inject

@PerGameMap
class SpawnSystem @Inject constructor(
    private val notificationController: NotificationController,
    private val entityProcessorSystem: EntityProcessorSystem,
    private val dataRepository: DataRepository
) : BaseSystem(Family.all(VisionComponent::class.java, BodyComponent::class.java).exclude(
    PassivityComponent::class.java).get()) {
    lateinit var spawns: ImmutableArray<Entity>
    private var groupEntities: ImmutableArray<Entity>? = null
    private val sm = ComponentMapper.getFor(SpawnComponent::class.java)
    private val pm = ComponentMapper.getFor(BodyComponent::class.java)
    private val gm = ComponentMapper.getFor(StalkerGroup::class.java)

    override fun addedToEngine(engine: Engine) {
        super.addedToEngine(engine)
        spawns = engine.getEntitiesFor(
            Family
                .all(SpawnComponent::class.java, BodyComponent::class.java)
                .exclude(PassivityComponent::class.java)
                .get()
        )
        val stalkerGroupFamily = Family.all(StalkerGroup::class.java, BodyComponent::class.java).get()

        groupEntities = engine.getEntitiesFor(stalkerGroupFamily)

        engine.addEntityListener(stalkerGroupFamily, object : EntityListener {
            override fun entityAdded(entity: Entity?) {
            }

            override fun entityRemoved(entity: Entity?) {
                gm[entity].removeEntity(entity)
            }

        })
    }

    var timer: Float = 0f
    var secsToTakeSpawn: Int = 10
    var takingSpawm: SpawnComponent? = null

    override fun update(deltaTime: Float) {
        super.update(deltaTime)
        for (i in 0 until spawns.size()) {
            val spawnComponent = sm[spawns[i]]
            val spawnBodyComponent = pm[spawns[i]]

            if (!spawnComponent.isEmpty)
                continue

            if (!spawnComponent.isActionsDone) {
                Gdx.app.applicationLogger.log("Spawn actions", "Actions sent")
                dataRepository.applyActions(spawnComponent.spawnModel.actions)
                spawnComponent.isActionsDone = true
            }


            var minDst = 120f
            val dstToPlayer = pm.get(player).position.dst(spawnBodyComponent.position)
            if (dstToPlayer < minDst) {
                if (takingSpawm == null) {
                    takingSpawm = spawnComponent
                    timer = secsToTakeSpawn.toFloat()
                }
                notificationController.setTitle("Захват позиции через " + timer.toInt() + " cекунд")
                timer-=deltaTime
                if (timer < 1) {
                    notificationController.setTitle("Позиция захвачена!")
                    notificationController.addMessage("Спасибо, уже отправили группу на занятие позиции.")
                    takingSpawm == null
                    val gang = dataRepository.initDataModel.gang
                    spawnComponent.stalkerGroup = entityProcessorSystem.generateNewGroup(gang)
                }
            } else if (takingSpawm == spawnComponent)
                takingSpawm = null
            else {
                var minDstStalkerGroup: StalkerGroup? = null
                for (entity in groupEntities!!) {
                    val group = gm[entity]
                    if (group.targeting !is SpawnComponent) {
                        val dst = group.centerPoint.dst(spawnBodyComponent.position)
                        if (minDstStalkerGroup == null || dst < minDst) {
                            minDstStalkerGroup = group
                            minDst = dst
                        }
                    }
                }
                if (minDstStalkerGroup != null)
                    spawnComponent.stalkerGroup = minDstStalkerGroup
            }
        }
    }

    override fun processEntity(entity: Entity, deltaTime: Float) {}

    val emptySpawn: Entity?
        get() {
            for (entity in spawns) {
                val spawnComponent = sm[entity]
                if (spawnComponent.isEmpty) {
                    return entity
                }
            }
            return null
        }

    val randomSpawn: Entity?
        get() = spawns.random()
}